### PR TITLE
Reduce allocations in `HttpRequestUtils.GetUrl()`

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

Reduces the allocations that occur in `HttpRequestUtils.GetUrl()`, particularly on .NET 6+

## Reason for change

While profiling a simple aspnetcore + httprequest app (our aspnetcore benchmark), this showed up as a _relatively_ high source of `string` allocations (6%) so wanted to try to reduce that. It was explicitly the version that takes a `Uri` that showed up, so optimized that somewhat.

## Implementation details

In newer runtimes (.NET 6+), we can use `GetComponents()` to avoid allocating each of the individual _pieces_ of url before formatting. We can also take advantage of the fact this whole path is only valid with absolute Uris, so we know we have a `Host`, and various other minor things.

> On older runtimes, `GetComponents()` _horrendously_ allocates, hence the split implementation 😅 

Otherwise, tried lots of things running various benchmarks 😅  `Uri` is a tricky beast to deal with, because it does a lot of internal caching too. In general though, we'll be dealing with a new `Uri` instance every time we call this method, so that's what I focused on. Benchmarked both approaches though, just to be sure, and we come out ahead in both cases.

<details><summary>Original benchmarks</summary>
<p>

| Method                 | Runtime            |      Mean |     Error | Allocated |
| ---------------------- | ------------------ | --------: | --------: | --------: |
| GetUrl_Original        | .NET 10.0          | 254.13 ns |  5.058 ns |     456 B |
| GetUrl_Updated         | .NET 10.0          | 215.42 ns |  4.340 ns |     288 B |
| GetUrl_Original        | .NET 6.0           | 384.78 ns |  4.289 ns |     440 B |
| GetUrl_Updated         | .NET 6.0           | 357.27 ns |  7.153 ns |     280 B |
| GetUrl_Original        | .NET Core 2.1      | 556.36 ns |  9.929 ns |     544 B |
| GetUrl_Updated         | .NET Core 2.1      | 495.81 ns |  7.232 ns |     472 B |
| GetUrl_Original        | .NET Core 3.1      | 546.39 ns |  7.438 ns |     528 B |
| GetUrl_Updated         | .NET Core 3.1      | 506.67 ns |  8.758 ns |     464 B |
| GetUrl_Original        | .NET Framework 4.8 | 727.18 ns | 10.738 ns |     786 B |
| GetUrl_Updated         | .NET Framework 4.8 | 783.27 ns | 43.809 ns |     706 B |
|                        |                    |           |           |           |
| GetUrl_Cached_Original | .NET 10.0          |  46.87 ns |  1.404 ns |      88 B |
| GetUrl_Cached_Updated  | .NET 10.0          |  45.29 ns |  0.871 ns |      88 B |
| GetUrl_Cached_Original | .NET 6.0           |  63.98 ns |  1.202 ns |      88 B |
| GetUrl_Cached_Updated  | .NET 6.0           |  61.60 ns |  1.099 ns |      88 B |
| GetUrl_Cached_Original | .NET Core 2.1      |  80.93 ns |  1.578 ns |     168 B |
| GetUrl_Cached_Updated  | .NET Core 2.1      |  69.74 ns |  1.411 ns |     160 B |
| GetUrl_Cached_Original | .NET Core 3.1      |  79.15 ns |  1.258 ns |     160 B |
| GetUrl_Cached_Updated  | .NET Core 3.1      |  67.23 ns |  1.314 ns |     152 B |
| GetUrl_Cached_Original | .NET Framework 4.8 |  84.30 ns |  1.350 ns |     241 B |
| GetUrl_Cached_Updated  | .NET Framework 4.8 |  66.33 ns |  0.709 ns |     225 B |

</p>
</details> 

EDIT: Shortly after submitting this PR, I discovered `DangerousDisablePathAndQueryCanonicalization` and tl;dr; that's a nightmare 😅 It's only in .NET 6+, but they don't expose whether it was set publicly, so had to do some gnarly stuff with duck types. I'm not proud, but the benchmarks looks worth it to me:

| Method                    | Runtime            |        Mean |      Error | Allocated |
| ------------------------- | ------------------ | ----------: | ---------: | --------: |
| GetUrl_Original           | .NET 10.0          | 253.9958 ns |  3.6006 ns |     456 B |
| GetUrl_Updated            | .NET 10.0          |   210.29 ns |   2.170 ns |     288 B |
| GetUrl_Original           | .NET 6.0           | 383.5104 ns |  2.6358 ns |     440 B |
| GetUrl_Updated            | .NET 6.0           | 352.9394 ns |  2.9057 ns |     280 B |
| GetUrl_Original           | .NET Core 2.1      | 541.4805 ns | 10.6203 ns |     544 B |
| GetUrl_Updated            | .NET Core 2.1      | 490.8919 ns |  5.8440 ns |     472 B |
| GetUrl_Original           | .NET Core 3.1      | 548.5983 ns | 10.8097 ns |     528 B |
| GetUrl_Updated            | .NET Core 3.1      | 505.9982 ns |  8.7285 ns |     464 B |
| GetUrl_Original           | .NET Framework 4.8 | 745.0813 ns | 14.5547 ns |     786 B |
| GetUrl_Updated            | .NET Framework 4.8 | 686.8337 ns |  2.9292 ns |     706 B |
|                           |                    |             |            |           |
| GetUrl_Dangerous_Original | .NET 10.0          | 208.3430 ns |  4.0426 ns |     488 B |
| GetUrl_Dangerous_Updated  | .NET 10.0          |   168.41 ns |   2.801 ns |     342 B |
| GetUrl_Dangerous_Original | .NET 6.0           | 319.3693 ns |  2.4285 ns |     472 B |
| GetUrl_Dangerous_Updated  | .NET 6.0           | 296.5951 ns |  2.0317 ns |     384 B |
|                           |                    |             |            |           |
| GetUrl_Redaction_Original | .NET 10.0          | 278.4834 ns |  3.6679 ns |     488 B |
| GetUrl_Redaction_Updated  | .NET 10.0          |   241.85 ns |   2.278 ns |     288 B |
| GetUrl_Redaction_Original | .NET 6.0           | 420.9738 ns |  4.4795 ns |     472 B |
| GetUrl_Redaction_Updated  | .NET 6.0           | 389.7784 ns |  3.2476 ns |     280 B |
| GetUrl_Redaction_Original | .NET Core 2.1      | 591.9719 ns |  6.4948 ns |     568 B |
| GetUrl_Redaction_Updated  | .NET Core 2.1      | 552.1795 ns |  7.9159 ns |     472 B |
| GetUrl_Redaction_Original | .NET Core 3.1      | 608.4211 ns |  7.3971 ns |     560 B |
| GetUrl_Redaction_Updated  | .NET Core 3.1      | 559.9308 ns |  6.6185 ns |     464 B |
| GetUrl_Redaction_Original | .NET Framework 4.8 | 818.2093 ns | 10.0129 ns |     834 B |
| GetUrl_Redaction_Updated  | .NET Framework 4.8 | 779.5907 ns |  7.6899 ns |     730 B |
|                           |                    |             |            |           |
| GetUrl_Cached_Original    | .NET 10.0          |  45.2706 ns |  0.5235 ns |      88 B |
| GetUrl_Cached_Updated     | .NET 10.0          |  40.9379 ns |  0.7165 ns |      88 B |
| GetUrl_Cached_Original    | .NET 6.0           |  59.6816 ns |  0.4011 ns |      88 B |
| GetUrl_Cached_Updated     | .NET 6.0           |  71.1780 ns |  1.2479 ns |      88 B |
| GetUrl_Cached_Original    | .NET Core 2.1      |  77.7640 ns |  1.0165 ns |     168 B |
| GetUrl_Cached_Updated     | .NET Core 2.1      |  63.4134 ns |  1.0179 ns |     160 B |
| GetUrl_Cached_Original    | .NET Core 3.1      |  72.5815 ns |  0.9315 ns |     160 B |
| GetUrl_Cached_Updated     | .NET Core 3.1      |  61.6694 ns |  0.6063 ns |     152 B |
| GetUrl_Cached_Original    | .NET Framework 4.8 |  86.4770 ns |  0.7504 ns |     241 B |
| GetUrl_Cached_Updated     | .NET Framework 4.8 |  72.9856 ns |  1.1003 ns |     225 B |

<details><summary>Benchmarking code</summary>
<p>

```csharp
 [MemoryDiagnoser, GroupBenchmarksBy(BenchmarkLogicalGroupRule.ByCategory), CategoriesColumn]
    public class HttpRequestUtilBenchmarks
    {
        private Uri _uri;
        private string _result;

        [GlobalSetup]
        public void GlobalSetup()
        {
            _uri = new Uri("https://api.example.com/v1/search?q=test&limit=10");
        }

        [Benchmark(Baseline = true)]
        [BenchmarkCategory("GetUrl", "Cached")]
        public int GetUrl_Cached_Original()
        {
            _result = OriginalHttpRequestUtils.GetUrl(_uri);
            return _result.Length;
        }

        [Benchmark]
        [BenchmarkCategory("GetUrl", "Cached")]
        public int GetUrl_Cached_Updated()
        {
            _result = HttpRequestUtils.GetUrl(_uri);
            return _result.Length;
        }

        [Benchmark(Baseline = true)]
        [BenchmarkCategory("GetUrl", "new")]
        public int GetUrl_Original()
        {
            var uri = new Uri("https://api.example.com/v1/search?q=test&limit=10");
            _result = OriginalHttpRequestUtils.GetUrl(uri);
            return _result.Length;
        }

        [Benchmark]
        [BenchmarkCategory("GetUrl", "new")]
        public int GetUrl_Updated()
        {
            var uri = new Uri("https://api.example.com/v1/search?q=test&limit=10");
            _result = HttpRequestUtils.GetUrl(uri);
            return _result.Length;
        }

        [Benchmark(Baseline = true)]
        [BenchmarkCategory("GetUrl", "redaction")]
        public int GetUrl_Redaction_Original()
        {
            var uri = new Uri("http://localhost/api?id=123&api_key=secret&sort=asc");
            _result = OriginalHttpRequestUtils.GetUrl(uri);
            return _result.Length;
        }

        [Benchmark]
        [BenchmarkCategory("GetUrl", "redaction")]
        public int GetUrl_Redaction_Updated()
        {
            var uri = new Uri("http://localhost/api?id=123&api_key=secret&sort=asc");
            _result = HttpRequestUtils.GetUrl(uri);
            return _result.Length;
        }

#if NET6_0_OR_GREATER
        [Benchmark(Baseline = true)]
        [BenchmarkCategory("GetUrl", "new", "dangerous")]
        public int GetUrl_Dangerous_Original()
        {
            var uri = new Uri("http://localhost/api?id=123&api_key=secret&sort=asc", new UriCreationOptions(){ DangerousDisablePathAndQueryCanonicalization = true});
            _result = OriginalHttpRequestUtils.GetUrl(uri);
            return _result.Length;
        }

        [Benchmark]
        [BenchmarkCategory("GetUrl", "new", "dangerous")]
        public int GetUrl_Dangerous_Updated()
        {
            var uri = new Uri("http://localhost/api?id=123&api_key=secret&sort=asc", new UriCreationOptions(){ DangerousDisablePathAndQueryCanonicalization = true});
            _result = HttpRequestUtils.GetUrl(uri);
            return _result.Length;
        }
#endif
```

</p>
</details> 

## Test coverage

Before making the change, 🤖 Claude created some characterization tests to avoid any regressions in behaviour.

## Other details

There's also `HttpRequestExtensions.GetUrlForSpan()` which has similar behaviour and allocations today, that I will likely investigate similarly

Spotted while working on https://datadoghq.atlassian.net/browse/LANGPLAT-842

### CODE REVIEW

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs [hunk 1]
```diff
@@ -15,15 +14,78 @@ internal static class HttpRequestUtils
     {
         private const string NoHostSpecified = "UNKNOWN_HOST";
 
+#if NET6_0_OR_GREATER
+        // In .NET 6+, we could theoretically bypass a bunch of allocations by using the GetComponents() method which is heavily
+        // optimized. Unfortunately, in .NET FX and < .NET 6, this approach allocates a _lot_ more. And what's more
+        // .NET 6+ introduces 'DangerousDisablePathAndQueryCanonicalization' which means calling GetComponents() _Throws_, and
+        // we have no way to detect it
```
Comment (lucaspimentel) [inline tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: "...and we have no way to detect it"

But we're detecting it below with duck-casting. Does the comment need updating?

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs [hunk 2]
```diff
@@ -37,5 +99,22 @@ internal static string GetUrl(string scheme, string host, int? port, string path
         }
 
         internal static string GetNormalizedHost(string? host) => StringUtil.IsNullOrEmpty(host) ? NoHostSpecified : host;
+
+#if NET6_0_OR_GREATER
+        [DuckCopy]
+        internal struct UriStruct
```
Comment (lucaspimentel) [inline tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: You're adding the same duck type in #8199 at `tracer/src/Datadog.Trace/Util/UriHelpers.cs`.

Is that intentional?

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs (hunk 1 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: I guess, it should probably be "no way to detect it with public APIs" 🤔

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs [hunk 3]
```diff
@@ -15,15 +14,78 @@ internal static class HttpRequestUtils
...
-                uri.Scheme,
-                uri.Host,
-                uri.IsDefaultPort ? null : uri.Port,
-                string.Empty,
-                uri.AbsolutePath,
-                uri.Query,
-                queryStringManager);
+            => uri.DuckCast<UriStruct>().IsDangerousDisablePathAndQueryCanonicalization()
+                   ? GetUrlSimple(uri, queryStringManager)
+                   : GetUrlDangerous(uri, queryStringManager);
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: Agreed, renamed them

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: That's not true, it throws

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs [hunk 4]
```diff
@@ -37,5 +99,22 @@ internal static string GetUrl(string scheme, string host, int? port, string path
...
+
+#if NET6_0_OR_GREATER
+        [DuckCopy]
+        internal struct UriStruct
+        {
+            // the flag changed in .NET 10
+            private static readonly ulong DisablePathAndQueryCanonicalizationFlag
+                = FrameworkDescription.Instance.RuntimeVersion.Major >= 10
+                      ? 1UL << 55
+                      : 0x200000000000;
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: Updated documentation, added link to original

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs (hunk 1 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: Updated the docs

File: tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs (hunk 2 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs]: It was, but I've removed the duplciation now 🙂

File: tracer/test/Datadog.Trace.Tests/Util/Http/HttpRequestUtilsTests.cs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    // Fragments are removed (by Uri class)
+    [InlineData("http://localhost/path#section", false, false, "http://localhost/path")]
+    [InlineData("http://localhost/path#section", true, false, "http://localhost/path")]
+    [InlineData("http://localhost/path?key=value#section", false, false, "http://localhost/path")]
+    [InlineData("http://localhost/path?key=value#section", true, false, "http://localhost/path?key=value")]
+
+    // With dangerous create. the fragment isn't removed
+    [InlineData("http://localhost/path#section", false, true, "http://localhost/path#section")]
+    [InlineData("http://localhost/path#section", true, true, "http://localhost/path#section")]
+    [InlineData("http://localhost/path?key=value#section", false, true, "http://localhost/path")]
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/test/Datadog.Trace.Tests/Util/Http/HttpRequestUtilsTests.cs:50]: I know, it's weird. It's because the section ends up in the querystring in this case 🤷 tl;dr; this is effectively a bug in our code, but it's such an edge case, I don't think it needs prioritizing tbh
